app/components/supplementary-file-browser.js

```diff
@@ -25,100 +27,95 @@ import fileDownloadPath from '../utils/file-download-path';
+                .then(this.__serializeVersions.bind(this, versions)),
+        });
     }),
-    _moveIfNeeded: observer('selectedFile', function() { /* eslint-disable-line ember/no-observers */
-        const index = this.get('files') && this.get('files').indexOf(this.get('selectedFile'));
-        if (index < 0) {
-            return;
-        }
-        if (index >= this.get('endIndex') || index < this.get('startIndex')) {
-            const max = this.get('files').length - 6;
-            if (index > max) {
-                this.set('startIndex', max);
-                this.set('endIndex', this.get('files').length);
-            } else {
-                this.set('startIndex', index);
-                this.set('endIndex', index + 6);
-            }
-        }
+
+    fileRendererURL: computed('selectedFile.links.download', 'selectedVersion', function() {
+        const downloadUrl = this.get('selectedFile.links.download');
+        const selectedVersion = this.get('selectedVersion');
+
+        return (downloadUrl && selectedVersion) ? `${downloadUrl}?version=${selectedVersion}` : null;
     }),
 
-    // This needs to be changed away from an observer, but "preprint" changes
-    // frequently enough that it is really complicated
-    __files: observer('preprint', function() { /* eslint-disable-line ember/no-observers */
-        this.set('files', []);
-        this.set('selectedFile', null);
-        /* eslint-disable ember/named-functions-in-promises */
-        this.get('node').get('files')
-            .then((fileProviders) => {
-                this.set('fileProvider', fileProviders.findBy('name', 'osfstorage'));
-                return loadAll(this.get('fileProvider'), 'files', this.get('files'), { 'page[size]': 50 });
-            })
-            .then(() => this.get('preprint').get('primaryFile'))
-            .then((pf) => {
-                this.get('files').removeObject(pf);
-                this.set('primaryFile', pf);
-                this.set('selectedFile', this.get('primaryFile'));
-                this.set('files', [this.get('primaryFile')].concat(this.get('files')));
-                this.set('indexes', this.get('files').map(each => each.id));
-            });
-        /* eslint-enable ember/named-functions-in-promises */
+    fileDownloadURL: computed('selectedFile', 'selectedVersion', function() {
+        return fileDownloadPath(this.get('selectedFile'), this.get('node'));
     }),
+
+    __files: function() {
+        this.get('node.files')
+            .then(providers => loadAll(
+                providers.findBy('name', 'osfstorage'),
+                'files',
+                this.get('files'),
+                {
+                    'page[size]': 50,
+                },
+            ))
+            .then(() => this.get('preprint.primaryFile'))
+            .then(this.__initProperties.bind(this));
+    }.observes('preprint'),
```

Comment:
I don't know how I should feel about using `observes` here. This is probably due to the fact that promises in computed functions don't work so well. But I have found some literature on [this](https://stackoverflow.com/questions/36945858/promise-result-in-ember-data-computed-property) which may worth a look.

Reply:
I feel like the `__files()` method could be refactored to a computed function while making use of `ember-concurrency`. I don't like the chaining promises and the `observes` in particular. I can see that this is somewhat a port of what was done in #285 . Therefore if we can refactor it with not much effort, it would be nice. Maybe don't worry about it if it takes too much time.